File: solutions/java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private final int speed;
    private final int batteryDrain;

    private int distance;
    private int batteryLevel;

    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
        this.distance = 0;
        this.batteryLevel = 100;
    }

    public boolean batteryDrained() {
        return batteryLevel < batteryDrain;
    }

    public int distanceDriven() {
        return distance;
    }

    public void drive() {
        if (!batteryDrained()) {
            distance += speed;
            batteryLevel -= batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private final int distance;

    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        while (car.distanceDriven() < distance) {
            if (car.batteryDrained())
                return false;
            car.drive();
        }
        return true;
    }
}

File: solutions/csharp/need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private int _distance = 0;
    private int _battery = 100;
    private int _speed = 20;
    private int _batteryDrain = 1;

    // Creating a remote controlled car
    public RemoteControlCar(int speed, int batteryDrain)
    {
        this._speed = speed;
        this._batteryDrain = batteryDrain;
    }

    // Check for a drained battery
    public bool BatteryDrained() => this._battery < this._batteryDrain;

    public int DistanceDriven() => this._distance;

    // Drive the car
    public void Drive()
    {
        if (!BatteryDrained())
        {
            _distance += _speed;
            _battery -= _batteryDrain;
        }
    }

    // Create the Nitro remote control car
    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int _distance;

    // Creating a race track
    public RaceTrack(int distance) => this._distance = distance;

    // Check if a remote control car can finish a race
    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained())
        {
            car.Drive();
            if (car.DistanceDriven() >= this._distance)
            {
                return true;
            }
        }
        return false;
    }
}
